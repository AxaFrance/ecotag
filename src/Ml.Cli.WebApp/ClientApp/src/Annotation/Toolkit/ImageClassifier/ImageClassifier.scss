.image-classifier{
  justify-content: center;
  text-align: center;
  width:100%;
  &__image-container{
    display: flex;
    justify-content: center;
    width:100%;
  }
  &__image{
    width: 80%;
  }
  &__buttons-container{
    display: flex;
    flex-wrap: wrap;
    justify-content: space-evenly;
    margin-bottom: 1em;
  }
  &__button-container{
    margin-top: 3em;
    flex: 1 0 21%;
  }
}
