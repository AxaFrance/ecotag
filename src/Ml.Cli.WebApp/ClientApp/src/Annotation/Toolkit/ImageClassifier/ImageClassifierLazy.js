import React, {Suspense} from "react";

const ImageClassifierLazy = ({url, labels, onSubmit}) => {
    
    const ImageClassifier = React.lazy(() => import('./ImageClassifier'));
    
    return <Suspense fallback={<div>Loading image classifier...</div>}>
        <ImageClassifier
            url={url}
            labels={labels}
            onSubmit={onSubmit}
        />
    </Suspense>
    
};

export default ImageClassifierLazy;
