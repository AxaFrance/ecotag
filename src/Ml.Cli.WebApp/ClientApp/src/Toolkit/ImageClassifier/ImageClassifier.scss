.image-classifier{
  justify-content: center;
  text-align: center;
  width:100%;
  &--inline-mode{
    display: flex;
  }
  &__image-container{
    display: flex;
    justify-content: center;
    width:100%;
    align-items: center;
  }
  &__image{
    width: 80%;
  }
  &__buttons-container{
    display: flex;
    flex-wrap: wrap;
    justify-content: space-evenly;
    margin-bottom: 1em;
    &--inline-mode{
      flex-direction: column;
      flex-wrap: nowrap;
      justify-content: flex-start;
    }
  }
  &__button-container{
    margin-top: 3em;
    flex: 1 0 21%;
    &--inline-mode{
      margin-top: 1em;
      flex: none;
    }
  }
}

.image-classifier-btn{
  bottom: 15px;
  border: none;
  color: #fff;
  font-weight: 700;
  border-radius: 5px;
  padding: 10px 25px;
  cursor: pointer;
  min-width: 150px;
  &:hover {
    opacity: 0.7;
    transition-duration: 200ms;
    transition: ease-in;
  }
  &--selected {
    border: 4px dashed black;
  }
}