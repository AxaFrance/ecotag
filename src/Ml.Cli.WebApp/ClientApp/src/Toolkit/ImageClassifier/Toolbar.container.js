import React from 'react';
import { GlobalHotKeys } from 'react-hotkeys';
import {ToolbarButtonContainer, ToolbarButton, ToolbarProgressButton, ToolbarSwitchButton} from '../Toolbar';

const ToolbarContainer = ({ state, setState }) => {
  const getImageInfo = () => {
    const image = document.getElementById('currentImage');
    const imageWidth = image.width;
    const imageHeight = image.height;
    return {
      imageWidth,
      imageHeight,
    };
  };

  const rotateImage = value => {
    const { imageWidth, imageHeight } = getImageInfo();
    const margin = (Math.max(imageWidth, imageHeight) - Math.min(imageWidth, imageHeight)) / 2;
    const orientation = imageWidth > imageHeight ? 'horizontal' : 'vertical';
    const construcMargin = orientation === 'horizontal' ? `${margin}px -${margin}px` : `-${margin}px ${margin}px`;
    setState({
      ...state,
      marginRotate: construcMargin,
      rotate: value === 'right' ? state.rotate + 90 : state.rotate - 90,
      initialRotate: !state.initialRotate,
    });
  };

  const inlineMode = () => {
    setState({ ...state, inlineMode: !state.inlineMode });
  };

  const keyMap = {
    Submit: 'ctrl+spacebar',
    RotateLeft: 'l',
    RotateRight: 'r',
    InlineMode: 'i',
  };

  const handlers = {
    RotateRight: () => rotateImage('right'),
    RotateLeft: rotateImage,
    InlineMode: inlineMode,
  };

  return (
    <div className="toolbar">
      <GlobalHotKeys allowChanges={true} keyMap={keyMap} handlers={handlers}>
          <ToolbarButtonContainer classModifier="progress-bar">
            <ToolbarProgressButton
              label="Image Size"
              onChange={e => setState({ ...state, widthImage: e.target.value })}
              value={state.widthImage}
              min="1"
              max="80"
            />
            <ToolbarSwitchButton id="inlineMode" checked={state.inlineMode} onChange={inlineMode} label="Inline mode" />
          </ToolbarButtonContainer>
          <ToolbarButtonContainer>
            <ToolbarButton title="Raccourci : L" onClick={rotateImage} icon="reset" label="Rotate Left" />
            <ToolbarButton
              title="Raccourci : R"
              onClick={() => rotateImage('right')}
              icon="repeat"
              label="Rotate Right"
            />
          </ToolbarButtonContainer>
      </GlobalHotKeys>
    </div>
  );
};

export default ToolbarContainer;
