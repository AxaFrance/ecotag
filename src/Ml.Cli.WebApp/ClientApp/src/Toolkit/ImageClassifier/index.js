export { default } from './ImageClassifier.container';
