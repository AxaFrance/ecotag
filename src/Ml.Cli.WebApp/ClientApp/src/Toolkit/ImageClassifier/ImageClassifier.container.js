import React, {useEffect, useState} from "react";
import Toolbar from "./Toolbar.container";
import '../Toolbar/Toolbar.scss';
import ImageClassifier from "./ImageClassifier";

const ImageClassifierContainer = ({url, labels, onSubmit, expectedOutput}) => {
    const [state, setState] = useState({
        labels,
        url,
        rotate: 0,
        widthImage: 80,
        inlineMode: false,
        marginRotate: 0,
        initialRotate: true,
        userInput: {},
    });

    useEffect(() => {
        const userInput = {};
        for (const [key, value] of Object.entries(labels)) {
            userInput[key] = value;
        }
        setState({ ...state, userInput });
    }, [url]);
    
    return (
        <>
            <ImageClassifier onSubmit={onSubmit} url={url} labels={labels} state={state} expectedOutput={expectedOutput}/>
            <Toolbar state={state} setState={setState}/>
        </>
    )
};

export default React.memo(ImageClassifierContainer);