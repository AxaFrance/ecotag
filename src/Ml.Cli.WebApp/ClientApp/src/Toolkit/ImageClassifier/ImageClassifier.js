import React from "react";
import Button from "@axa-fr/react-toolkit-button";
import stringToRGB from "./color";
import '../BoundingBox/Labels.scss';
import './ImageClassifier.scss';
import '@axa-fr/react-toolkit-button/src/button.scss'
import classNames from "classnames";

const defaultClassName = 'image-classifier';
const defaultClassNameButtonsContainer = 'image-classifier__buttons-container';
const defaultClassNameButtonContainer = 'image-classifier__button-container';

const ImageClassifier = ({url, labels, onSubmit, state}) => {
    const className = classNames(defaultClassName, {
        [`${defaultClassName}--inline-mode`]: state.inlineMode,
    });
    const classNameButtonsContainer = classNames(defaultClassNameButtonsContainer, {
        [`${defaultClassNameButtonsContainer}--inline-mode`]: state.inlineMode,
    });
    const classNameButtonContainer = classNames(defaultClassNameButtonContainer, {
        [`${defaultClassNameButtonContainer}--inline-mode`]: state.inlineMode,
    });
    const coloredLabels = labels.map((label) => {
        return {
            "name": label.name,
            "color": `#${stringToRGB(label.name)}`
        };
    });
    
    return(
        <div className={className}>
            <div className={classNameButtonsContainer}>
                {coloredLabels.map((label, index) => {
                    return(
                        <div key={index} className={classNameButtonContainer}>
                            <Button onClick={() => onSubmit(label.name)} style={{backgroundColor: label.color, boxShadow: "none"}}>{label.name}</Button>
                        </div>
                    );
                })}
            </div>
            <div className="image-classifier__image-container">
                <img
                    src={url}
                    id="currentImage"
                    alt="Classifier image"
                    style={{
                        width: `${state.widthImage}%`,
                        transform: `rotate(${state.rotate}deg)`,
                        margin: `${state.initialRotate ? '' : state.marginRotate}`,
                    }}
                />
            </div>
        </div>
    );
};

export default ImageClassifier;