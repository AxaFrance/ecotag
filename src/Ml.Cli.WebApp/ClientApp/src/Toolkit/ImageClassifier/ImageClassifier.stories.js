import React from "react";
import ImageClassifier from "./ImageClassifier";
import url from "./sample_image.png";
import ImageClassifierContainer from "./ImageClassifier.container";

const mockedFunction = () => {};

const labels = [{name: "Dog"}, {name: "Cat"}, {name: "Duck"}, {name: "Other"}];

export default {
    title: 'ImageClassifier',
    component: ImageClassifier
}

const Template = (args) => <ImageClassifierContainer {...args}/>;

export const Default = Template.bind({});
Default.args = {
    url: url,
    labels: labels,
    onSubmit: mockedFunction
};