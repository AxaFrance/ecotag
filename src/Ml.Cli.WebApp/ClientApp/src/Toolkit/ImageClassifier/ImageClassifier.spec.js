import ImageClassifier from "./ImageClassifier";
import React from "react";
import {render, waitFor, fireEvent} from '@testing-library/react';

const labels = [{name: "Dog"}, {name: "Cat"}, {name: "Duck"}, {name: "Other"}];
import url from './sample_image.png';

describe("Check Image Classifier behaviour", () => {
   test("Should render correctly and click on button", async () => {
       
       let selectedValue = "";
       
       const {container, getByText, asFragment} = render(
           <ImageClassifier
               url={url}
               labels={labels}
               onSubmit={(e) => selectedValue = e}
           />
       )
       await waitFor(() => expect(container.querySelector(".image-classifier__buttons-container")).not.toBeNull());
       expect(asFragment()).toMatchSnapshot();
       
       const buttonsNumber = container.querySelectorAll(".af-btn");
       
       expect(buttonsNumber.length).toEqual(4);
       
       const selectedButton = getByText("Cat");
       fireEvent.click(selectedButton);
       
       expect(selectedValue).toEqual("Cat");
   }) ;
});