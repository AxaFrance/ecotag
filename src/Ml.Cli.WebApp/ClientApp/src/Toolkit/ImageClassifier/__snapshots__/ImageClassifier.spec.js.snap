// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Image Classifier behaviour Should render correctly and click on button 1`] = `
<DocumentFragment>
  <div
    class="image-classifier"
  >
    <div
      class="image-classifier__buttons-container"
    >
      <div
        class="image-classifier__button-container"
      >
        <button
          class="image-classifier-btn"
          style="background-color: rgb(1, 13, 28);"
        >
          Dog
        </button>
      </div>
      <div
        class="image-classifier__button-container"
      >
        <button
          class="image-classifier-btn"
          style="background-color: rgb(1, 7, 182);"
        >
          Cat
        </button>
      </div>
      <div
        class="image-classifier__button-container"
      >
        <button
          class="image-classifier-btn"
          style="background-color: rgb(32, 172, 217);"
        >
          Duck
        </button>
      </div>
      <div
        class="image-classifier__button-container"
      >
        <button
          class="image-classifier-btn"
          style="background-color: rgb(143, 142, 240);"
        >
          Other
        </button>
      </div>
    </div>
    <div
      class="image-classifier__image-container"
    >
      <img
        alt="Classifier image"
        id="currentImage"
        src="sample_image.png"
        style="width: 80%; transform: rotate(0deg);"
      />
    </div>
  </div>
  <div
    class="toolbar"
  >
    <div
      class="toolbar__button-container toolbar__button-container--progress-bar"
    >
      <div
        class="toolbar__button-progress-item"
      >
        <input
          class="toolbar__button-progress-item-input"
          max="80"
          min="1"
          name="width"
          type="range"
          value="80"
        />
        <label
          class="toolbar__button-progress-item-label"
        >
          Image Size
        </label>
      </div>
      <div
        class="toolbar__button-switch-item"
      >
        <input
          class="toolbar__button-switch-item-input"
          id="inlineMode"
          type="checkbox"
        />
        <label
          class="toolbar__button-switch-item-label"
          for="inlineMode"
        >
          Inline mode
        </label>
      </div>
    </div>
    <div
      class="toolbar__button-container"
    >
      <button
        class="toolbar__button-item"
        title="Raccourci : L"
      >
        <span
          class="glyphicon glyphicon-reset"
        />
        <div
          class="toolbar__button-item-label"
        >
          Rotate Left
        </div>
      </button>
      <button
        class="toolbar__button-item"
        title="Raccourci : R"
      >
        <span
          class="glyphicon glyphicon-repeat"
        />
        <div
          class="toolbar__button-item-label"
        >
          Rotate Right
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
